pkg/scalers/etcd_scaler.go

```diff
@@ -0,0 +1,306 @@
+func (s *etcdScaler) GetMetricSpecForScaling(context.Context) []v2.MetricSpec {
+	externalMetric := &v2.ExternalMetricSource{
+		Metric: v2.MetricIdentifier{
+			Name: GenerateMetricNameWithIndex(s.metadata.scalerIndex, kedautil.NormalizeString(fmt.Sprintf("etcd-%s-%s", s.metadata.etcdEndpoints, s.metadata.etcdKey))),
+		},
+		Target: GetMetricTargetMili(s.metricType, s.metadata.value),
+	}
+	metricSpec := v2.MetricSpec{External: externalMetric, Type: etcdMetricType}
+	return []v2.MetricSpec{metricSpec}
+}
+
+func (s *etcdScaler) Run(ctx context.Context, active chan<- bool) {
+	defer close(active)
+
+	// It's possible for the watch to get terminated anytime, we need to run this in a retry loop
+	runWithWatch := func() {
+		s.logger.Info("run watch", "etcdKey", s.metadata.etcdKey, "etcdEndpoints", s.metadata.etcdEndpoints)
+		subCtx, cancel := context.WithCancel(ctx)
+		subCtx = clientv3.WithRequireLeader(subCtx)
+		rch := s.client.Watch(subCtx, s.metadata.etcdKey, clientv3.WithProgressNotify())
+
+		// rewatch to another etcd server when the network is isolated from the current etcd server.
+		progress := make(chan bool)
+		defer close(progress)
+		go func() {
+			for {
+				select {
+				case <-progress:
+				case <-subCtx.Done():
+					return
+				case <-time.After(time.Duration(s.metadata.watchProgressNotifyInterval) * 2 * time.Second):
+					s.logger.Info("no watch progress notification in the interval", "etcdKey", s.metadata.etcdKey, "etcdEndpoints", s.metadata.etcdEndpoints)
+					cancel()
+					return
+				}
+			}
+		}()
+
+		for wresp := range rch {
+			progress <- wresp.IsProgressNotify()
+
+			// rewatch to another etcd server when there is an error form the current etcd server, such as 'no leader','required revision has been compacted'
+			if wresp.Err() != nil {
+				s.logger.Error(wresp.Err(), "an error occurred in the watch process", "etcdKey", s.metadata.etcdKey, "etcdEndpoints", s.metadata.etcdEndpoints)
+				cancel()
+				return
+			}
+
+			for _, ev := range wresp.Events {
+				v, err := strconv.ParseFloat(string(ev.Kv.Value), 64)
+				if err != nil {
+					s.logger.Error(err, "etcdValue invalid will be treated as 0")
+					v = 0
+				}
+				fmt.Println(v)
+				active <- v > s.metadata.activationValue
+			}
+		}
+	}
+
+	// retry on error from runWithWatch() starting by 2 sec backing off * 2 with a max of 1 minute
+	retryDuration := time.Second * 2
+	// the caller of this function needs to ensure that they call Stop() on the resulting
+	// timer, to release background resources.
+	retryBackoff := func() *time.Timer {
+		tmr := time.NewTimer(retryDuration)
+		retryDuration *= 2
+		if retryDuration > time.Minute*1 {
+			retryDuration = time.Minute * 1
+		}
+		return tmr
+	}
+
+	// start the first runWithWatch without delay
+	runWithWatch()
+
+	for {
+		backoffTimer := retryBackoff()
+		select {
+		case <-ctx.Done():
+			backoffTimer.Stop()
+			return
+		case <-backoffTimer.C:
+			backoffTimer.Stop()
+			runWithWatch()
+		}
+	}
+}
```

Comment:
I can't see where are we calling to this function. Are we using it?

Reply:
Yes, we use it in PushScaler interface. 
PushScaler combines the normal scaler interface and starts with the scale loop in the HandleScalableObject function.
The relevant code is as follows:
`go h.startPushScalers(ctx, withTriggers, obj.DeepCopy(), scalingMutex)`
`go h.startScaleLoop(ctx, withTriggers, obj.DeepCopy(), scalingMutex)`

Reply:
true, I have just seen it, thanks! 🙇